File: Classes/HSCalendarPasswordController.rb
####################################
#
#  HSCalendarPasswordController.rb
#
#  Author: Igor Sales.
#  Copyright (c) 2011 Igor Sales. All rights reserved.
#

# As found on http://bjeanes.com/2008/04/10/rubycocoa-and-keychain-access

framework 'Security'

class HSCalendarPasswordControllerVoid
    
    SERVICE = 'HoursCalendarPassword'
    
    def self.serviceName
        SERVICE
    end
    
    def self.passwordForUsername(username)
        service = serviceName

        status, *password = SecKeychainFindGenericPassword(nil, service.length, service, username.length, username)
        
        # Password-related data
        password_length = password.shift
        password_data   = password.shift # OSX::ObjcPtr object
        password        = password_data.bytestr(password_length)
    end

    def self.setPassword(password, forUsername: username)
        service = serviceName

        SecKeychainAddGenericPassword(nil, service.length, service, username.length, username, password.length, 
                                      password, nil)
    end
end

File: Classes/HSCalendarPasswordController.m
//
//  HSCalendarPasswordController.m
//  Hours
//
//  Created by Igor Sales on 12-03-14.
//  Copyright (c) 2012 igorsales.ca. All rights reserved.
//

#import "HSCalendarPasswordController.h"
#import <Security/Security.h>

#define SERVICE  @"HoursCalendarPassword"


@implementation HSCalendarPasswordController


+ (NSString*)serviceName
{
    return SERVICE;
}

+ (NSString*)passwordForUsername:(NSString*)username
{
    NSString* service = self.serviceName;

    UInt32 passwordLength = 0;
    const char* passwordData = nil;
    SecKeychainFindGenericPassword(nil, (UInt32)service.length, [service UTF8String], 
                                        (UInt32)username.length, [username UTF8String],
                                        &passwordLength, (void**)&passwordData, nil);
    
    NSString* password = nil;
    if (passwordData && passwordLength) {
        // It's possible the buffer is not null terminated, so copy it, use it, and throw it away.
        char *buffer = malloc(passwordLength+1);
        memcpy(buffer, passwordData, passwordLength);
        buffer[passwordLength] = 0;
        password = [NSString stringWithCString:buffer encoding:NSUTF8StringEncoding];
        free(buffer);
        
        SecKeychainItemFreeContent(nil, (void*)passwordData);
    }

    return password;
}

+ (void)setPassword:(NSString*)password forUsername:(NSString*)username
{
    NSString* service = self.serviceName;

    SecKeychainAddGenericPassword(nil, (UInt32)service.length, [service UTF8String], 
                                       (UInt32)username.length, [username UTF8String], 
                                       (UInt32)password.length, [password UTF8String], nil);
}

@end
